Add support for pmem in Redis configuration

michalbiesek opened the pull request:
Should be merged after #120 

Details:
- config is handled by pmem-str-mode param in redis.conf
- string and embedded string structures are covered
- scenarios assumed that PMEM placement for value is mandatory

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/redis/121)
<!-- Reviewable:end -->

kasiawasiuta reviewed (commented):
> Reviewed 9 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/121)* status: all files reviewed, 3 unresolved discussions (waiting on @jschmieg, @KFilipek, and @michalbiesek)

___
*[redis.conf, line 1706 at r1](https://reviewable.io/reviews/pmem/redis/121#-LzqmJJ6A_l3k-Fea9Fh:-LzqmJJ6A_l3k-Fea9Fi:bsqujq0) ([raw file](https://github.com/pmem/redis/blob/33b78e5cf8b4ba85c44382a0cb0006c39ab3a252/redis.conf#L1706)):*
> ```
> ############################### PMEM CONFIG ###############################
> # Persistent memory allocation strategy for strings and embedded strings
> # pmem-str-mode none
> ```

What can be other options? 'strategy' is not descriptive enough imo
___
*[src/defrag.c, line 81 at r1](https://reviewable.io/reviews/pmem/redis/121#-LzqnISwA9T1QBV79sDS:-LzqnISwA9T1QBV79sDT:b-yv2pc1) ([raw file](https://github.com/pmem/redis/blob/33b78e5cf8b4ba85c44382a0cb0006c39ab3a252/src/defrag.c#L81)):*
> ```
> NULL,ptr
> ```
pls add whitespace
___
*[src/server.h, line 347 at r1](https://reviewable.io/reviews/pmem/redis/121#-LzqnsO_BemeUh0SnjfJ:-LzqnsO_BemeUh0SnjfK:ba7ogas) ([raw file](https://github.com/pmem/redis/blob/33b78e5cf8b4ba85c44382a0cb0006c39ab3a252/src/server.h#L347)):*
> ```
> after we return single threaded 
> ```
Could you please explain? 


<!-- Sent from Reviewable.io -->

Commit:
Add string support of PMEM configuration

- config is handled by pmem-str-mode param in redis.conf
- string and embedded string structures are covered
- scenarios assumed that PMEM placement for structure value is mandatory

michalbiesek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/121)* status: 7 of 9 files reviewed, 3 unresolved discussions (waiting on @jschmieg, @kasiawasiuta, and @KFilipek)

___
*[redis.conf, line 1706 at r1](https://reviewable.io/reviews/pmem/redis/121#-LzqmJJ6A_l3k-Fea9Fh:-LzwF4Kq-xdt4Sm4iCmU:b8qcwc) ([raw file](https://github.com/pmem/redis/blob/33b78e5cf8b4ba85c44382a0cb0006c39ab3a252/redis.conf#L1706)):*
<details><summary><i>Previously, kasiawasiuta wrote…</i></summary><blockquote>

What can be other options? 'strategy' is not descriptive enough imo
</blockquote></details>

The description of each strategy is placed in server.h
___
*[src/defrag.c, line 81 at r1](https://reviewable.io/reviews/pmem/redis/121#-LzqnISwA9T1QBV79sDS:-LzwF-aY6V1njdfn2UXT:b-896fix) ([raw file](https://github.com/pmem/redis/blob/33b78e5cf8b4ba85c44382a0cb0006c39ab3a252/src/defrag.c#L81)):*
<details><summary><i>Previously, kasiawasiuta wrote…</i></summary><blockquote>

> ```
> NULL,ptr
> ```
pls add whitespace
</blockquote></details>

Done.
___
*[src/server.h, line 347 at r1](https://reviewable.io/reviews/pmem/redis/121#-LzqnsO_BemeUh0SnjfJ:-LzwHTkjCmEJSX8EyBLT:b-g8uk) ([raw file](https://github.com/pmem/redis/blob/33b78e5cf8b4ba85c44382a0cb0006c39ab3a252/src/server.h#L347)):*
<details><summary><i>Previously, kasiawasiuta wrote…</i></summary><blockquote>

> ```
> after we return single threaded 
> ```
Could you please explain? 
</blockquote></details>

:) removed


<!-- Sent from Reviewable.io -->

KFilipek reviewed (changes requested):
> Reviewed 2 of 9 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/121)* status: all files reviewed, 4 unresolved discussions (waiting on @jschmieg, @kasiawasiuta, and @michalbiesek)

___
*[src/server.h, line 369 at r2](https://reviewable.io/reviews/pmem/redis/121#-M-9uff18HYf-UNRnfGV:-M-9uff18HYf-UNRnfGW:b-z8i1lt) ([raw file](https://github.com/pmem/redis/blob/ea9ce4a286e7397a8f8682f83041ed00b98fca8c/src/server.h#L369)):*
<details><summary><i alt="summary">Quoted 9 lines of code…</i></summary>

> ```
> #define PMEM_EMBSTR_ROBJVAL (1<<4) /* Embedded string: Robj + value on PMEM */
> #define PMEM_EMBSTR_KEY (1<<5) /* Embedded string: key on PMEM */
> #define PMEM_EMBSTR_DICTENTRY (1<<6) /* Embedded string: DictEntry on PMEM */
> #define PMEM_EMBSTR_ROBJVAL_KEY \
>      (PMEM_EMBSTR_ROBJVAL | PMEM_EMBSTR_KEY) /* Embedded string: Robj + value and key on PMEM */
> #define PMEM_EMBSTR_ROBJVAL_DICTENTRY \
>      (PMEM_EMBSTR_ROBJVAL | PMEM_EMBSTR_DICTENTRY) /* Embedded string: Robj + value and DictEntry on PMEM */
> #define PMEM_EMBSTR_ALL \
>      (PMEM_EMBSTR_ROBJVAL | PMEM_EMBSTR_KEY | PMEM_EMBSTR_DICTENTRY) /* Embedded string: Robj + value, key and DictEntry on PMEM */
> ```
</details>

Sometimes Robj, DictEntry starts from capital letter, sometimes not


<!-- Sent from Reviewable.io -->

michalbiesek reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/121)* status: 2 of 3 files reviewed, 4 unresolved discussions (waiting on @jschmieg, @kasiawasiuta, and @KFilipek)

___
*[src/server.h, line 369 at r2](https://reviewable.io/reviews/pmem/redis/121#-M-9uff18HYf-UNRnfGV:-M-9wc986dgPCMT0puZL:b-896fix) ([raw file](https://github.com/pmem/redis/blob/ea9ce4a286e7397a8f8682f83041ed00b98fca8c/src/server.h#L369)):*
<details><summary><i>Previously, KFilipek (Krzysztof Filipek) wrote…</i></summary><blockquote>

> ```
> #define PMEM_EMBSTR_ROBJVAL (1<<4) /* Embedded string: Robj + value on PMEM */
> #define PMEM_EMBSTR_KEY (1<<5) /* Embedded string: key on PMEM */
> #define PMEM_EMBSTR_DICTENTRY (1<<6) /* Embedded string: DictEntry on PMEM */
> #define PMEM_EMBSTR_ROBJVAL_KEY \
>      (PMEM_EMBSTR_ROBJVAL | PMEM_EMBSTR_KEY) /* Embedded string: Robj + value and key on PMEM */
> #define PMEM_EMBSTR_ROBJVAL_DICTENTRY \
>      (PMEM_EMBSTR_ROBJVAL | PMEM_EMBSTR_DICTENTRY) /* Embedded string: Robj + value and DictEntry on PMEM */
> #define PMEM_EMBSTR_ALL \
>      (PMEM_EMBSTR_ROBJVAL | PMEM_EMBSTR_KEY | PMEM_EMBSTR_DICTENTRY) /* Embedded string: Robj + value, key and DictEntry on PMEM */
> ```
Sometimes Robj, DictEntry starts from capital letter, sometimes not
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

KFilipek reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r3.
> *[Reviewable](https://reviewable.io/reviews/pmem/redis/121#-:-M-A-zLp29xY3oiPm8mL:bnfp4nl)* status: all files reviewed, 3 unresolved discussions (waiting on @jschmieg and @kasiawasiuta)



<!-- Sent from Reviewable.io -->
